### PR TITLE
Implement update checker for Jib CLI

### PR DESCRIPTION
**Pull request by mpeddada1**
For #3161

### CODE REVIEW

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/Build.java [hunk 1]
```diff
@@ -106,9 +109,15 @@ public Integer call() {
...
             spec.commandLine().getErr(),
             executor);
+    Future<Optional<String>> updateCheckFuture = Futures.immediateFuture(Optional.empty());
     try {
       JibCli.configureHttpLogging(commonCliOptions.getHttpTrace().toJulLevel());
-
+      GlobalConfig globalConfig = GlobalConfig.readConfig();
+      updateCheckFuture =
+          JibCli.newUpdateChecker(
+              globalConfig,
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/Build.java:118]:
```suggestion
              GlobalConfig.readConfig(),
```

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java [hunk 2]
```diff
@@ -67,6 +79,45 @@ static void logTerminatingException(
             + "\u001B[0m");
   }
 
+  static Future<Optional<String>> newUpdateChecker(
+      GlobalConfig globalConfig, Verbosity verbosity, Consumer<LogEvent> log) {
+    if (!verbosity.equals(Verbosity.info) || globalConfig.isDisableUpdateCheck()) {
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java:84]: I think when the verbosity >= info.

**Review by chanseokoh — COMMENTED** [review]
There's an issue with the update message that `UpdateChecker.performUpdateCheck()` generates. It says "... Update your build configuration to use the latest features and fixes!"

And I see that Jib build plugins and Jib CLI will write to the same `lastUpdateCheck` file, but I think that's fine.

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java [hunk 3]
```diff
@@ -67,6 +79,45 @@ static void logTerminatingException(
...
+      return Futures.immediateFuture(Optional.empty());
+    }
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    try {
+      return UpdateChecker.checkForUpdate(
+          executorService,
+          VERSION_URL,
+          JibCli.class.getPackage().getImplementationTitle(),
+          JibCli.class.getPackage().getImplementationVersion(),
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java:93]: So `.getPackage()` works only when the `JibCli` is packaged in a `.jar` file, which is true now. But I fear it may not work when we generate a native executable. Should we use `VersionInfo.TOOL_NAME` and `VersionInfo.getSimpleVersion()`?

File: jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java [hunk 4]
```diff
@@ -19,21 +19,37 @@
...
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
 
+@RunWith(MockitoJUnitRunner.class)
 public class JibCliTest {
 
+  @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  @Mock GlobalConfig globalConfig;
+  @Mock ConsoleLogger logger;
```
Comment (chanseokoh) [inline jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java:51]: private

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java [hunk 5]
```diff
@@ -122,11 +122,13 @@
...
         return Optional.of(
-            "A new version of Jib ("
+            "A new version of "
+                + toolName
+                + " ("
                 + version.latest
                 + ") is available (currently using "
                 + currentVersion
-                + "). Update your build configuration to use the latest features and fixes!");
+                + "). Upgrade to use the latest features and fixes!");
```
Comment (chanseokoh) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java:131]: So I think we want to customize the update message. For Jib CLI, it'll really help if it can directly present the GitHub Release for the new version.

**Comment by chanseokoh** [comment]
Hmm... I always thought there should be a way to refactor `UpdateChecker` and the code around it to be simpler and more general. I feel the current code is a bit convoluted.

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java (hunk 3 above)
Comment (mpeddada1) [reply jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java]: Good thinking

File: jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java [hunk 6]
```diff
@@ -68,4 +82,12 @@ public void testLogTerminatingException_stackTrace() {
         .log(LogEvent.Level.ERROR, "\u001B[31;1mjava.io.IOException: test error message\u001B[0m");
     verifyNoMoreInteractions(logger);
   }
+
+  @Test
+  public void testNewUpdateChecker_noUpdateCheck() throws ExecutionException, InterruptedException {
+    when(globalConfig.isDisableUpdateCheck()).thenReturn(true);
+    Future<Optional<String>> updateChecker =
+        JibCli.newUpdateChecker(globalConfig, Verbosity.info, ignored -> {});
+    assertThat(updateChecker.get()).isEqualTo(Optional.empty());
```
Comment (chanseokoh) [inline jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java:91]:
```suggestion
    assertThat(updateChecker.get()).isEmpty();
```

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java [hunk 7]
```diff
@@ -121,12 +121,22 @@
...
+        String updateMessage =
+            "A new version of "
+                + toolName
+                + " ("
                 + version.latest
                 + ") is available (currently using "
                 + currentVersion
-                + "). Update your build configuration to use the latest features and fixes!");
+                + "). ";
+        if (toolName.equals("jib-cli")) {
```
Comment (chanseokoh) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java:132]: Hmm... although it works, I think this is too hacky to accept. Maybe we can make this method return only the latest version string (i.e., `version.latest`), and have the caller construct the message (maybe in `finishUpdateChecker`)? Or, any other thoughts?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java (hunk 5 above)
Comment (mpeddada1) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java]: That does make sense. I modified the message for the jib-cli to include a link to the release page.

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java [hunk 8]
```diff
@@ -121,12 +121,22 @@
...
+          return Optional.of(
+              updateMessage
+                  + "Download the latest jib-cli version from https://github.com/GoogleContainerTools/jib/releases.");
```
Comment (chanseokoh) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java:135]: And since we have the `version.latest`, I think we can directly link the Release with the target version?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java (hunk 7 above)
Comment (mpeddada1) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java]: You're absolutely right. I refactored the code to make this method return `version.latest` and constructed the message in `finishUpdateCheck`.

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java (hunk 8 above)
Comment (mpeddada1) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java]: Good point

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java [hunk 9]
```diff
@@ -37,6 +48,9 @@
     subcommands = {Build.class, Jar.class})
 public class JibCli {
 
+  private static final String VERSION_URL = "https://storage.googleapis.com/jib-versions/jib-cli";
+  private static final String TOOL_VERSION = VersionInfo.getVersionSimple();
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java:52]: I find no need to declare it as a constant field.

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java [hunk 10]
```diff
@@ -67,6 +81,49 @@ static void logTerminatingException(
...
+                      + ") is available (currently using "
+                      + TOOL_VERSION
+                      + "). Download the latest jib-cli version from "
+                      + ProjectInfo.GITHUB_URL
+                      + "/releases/tag/v"
+                      + latestVersion
+                      + "-cli"
+                      + "\n"
+                      + ProjectInfo.GITHUB_URL
+                      + "/blob/master/jib-cli/CHANGELOG.md\u001B[0m"
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/JibCli.java:120]: I think we should have `\n` after this. Maybe `\n\n` is better?

It's a bit hard to parse this and check if this is correctly formed. Maybe `String.format` can improve readability? And I think we can just embed "Jib CLI" without `VersionInfo.TOOL_NAME`.

```java
String message = "A new version of Jib CLI (%d) is available (currently using %d). Download the latest Jib CLI from %s/release/tag/v%d-cli\n"
    + "%s/blob/master/jib-cli/CHANGELOG.md\u001B[0m\n\n"
    + ...
```

File: jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java [hunk 11]
```diff
@@ -17,23 +17,42 @@
...
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
 
+@RunWith(MockitoJUnitRunner.class)
 public class JibCliTest {
 
+  @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
Comment (chanseokoh) [inline jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java:51]: Why do we need this?

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TaskCommonTest.java [hunk 12]
```diff
@@ -71,7 +82,26 @@ public void testGetBootWarTask_bootWarProject() {
     project.getPlugins().apply(SpringBootPlugin.class);
 
     TaskProvider<Task> bootWarTask = TaskCommon.getBootWarTaskProvider(project);
-    Assert.assertNotNull(bootWarTask);
-    Assert.assertNotNull(bootWarTask instanceof BootWar);
+    assertThat(bootWarTask).isNotNull();
+    assertThat(bootWarTask instanceof BootWar).isNotNull();
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TaskCommonTest.java:86]: Doesn't it have `assertThat(bootWarTask).isInstanceOf(...)`?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java [hunk 13]
```diff
@@ -140,9 +134,9 @@
...
   /**
-   * Returns a message indicating Jib should be upgraded if the check succeeded and the current
-   * version is outdated, or returns {@code Optional.empty()} if the check was interrupted or did
-   * not determine that a later version was available.
+   * Returns the latest Jib version available if the check succeeded and the current version is
+   * outdated, or returns {@code Optional.empty()} if the check was interrupted or did not determine
+   * that a later version was available.
    *
    * @param updateMessageFuture the {@link Future} returned by {@link UpdateChecker#checkForUpdate}
    * @return the {@link Optional} message to upgrade Jib if a later version was found, else {@code
```
Comment (chanseokoh) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java:142]:
```suggestion
   * @return the latest version, if found
```

**Comment by chanseokoh** [comment]
And did you say you'll update CHANGELOG?

File: jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java (hunk 11 above)
Comment (mpeddada1) [reply jib-cli/src/test/java/com/google/cloud/tools/jib/cli/JibCliTest.java]: Good catch

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TaskCommonTest.java [hunk 14]
```diff
@@ -60,8 +71,8 @@ public void testGetWarTask_normalWarProject() {
     project.getPlugins().apply(WarPlugin.class);
 
     TaskProvider<Task> warTask = TaskCommon.getWarTaskProvider(project);
-    Assert.assertNotNull(warTask);
-    Assert.assertNotNull(warTask instanceof War);
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/TaskCommonTest.java:64]: Oh, looks like the old check was plainly wrong.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TaskCommon.java [hunk 15]
```diff
@@ -67,19 +68,27 @@
     }
   }
 
+  @VisibleForTesting
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TaskCommon.java:71]: The method is required to be package-private, not just for testing.
```suggestion
```

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MojoCommon.java [hunk 16]
```diff
@@ -76,18 +76,21 @@ static void finishUpdateChecker(
...
-                  LogEvent.lifecycle(
-                      "Please see https://github.com/GoogleContainerTools/jib/blob/master/docs/privacy.md for info on disabling this update check."));
-              projectProperties.log(LogEvent.lifecycle(""));
+            latestVersion -> {
+              String changelogUrl =
+                  ProjectInfo.GITHUB_URL + "/blob/master/jib-maven-plugin/CHANGELOG.md";
+              String privacyUrl = ProjectInfo.GITHUB_URL + "/blob/master/docs/privacy.md";
+              String message =
+                  String.format(
+                      "\n\u001B[33mA new version of %s (%s) is available (currently using %s). Update your"
```
Comment (chanseokoh) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MojoCommon.java:85]: Sorry, missed this. For Maven, we intentionally use multiple `.log()` instead of `\n` when we need to start a new line. Not a big deal, but this is because Maven prepends the verbosity info on every line, for example,
```
[INFO] Containerizing application to francium25/test...
```

**Review by chanseokoh — COMMENTED** [review]
LGTM. Only minor comments.
